### PR TITLE
Implemented Find Row(s) in the Data View Grid and fixed bug in the Edit Cell dialog

### LINKED ISSUE #7950: Add a simple Find dialogue into the Edit menu?
I would like to see the Edit menu "completed".  I am including @Patowhiz in this issue, just to get his take on the suggestions below.  Also it may prompt him to advise on how the Copy > Paste features can be completed - I think there is still a bit to go here.  Then @N-thony could advise, but this might be a simple task for Derrick or Victor, once agreed.

I would like to be able to use reogrid a bit more as it is possible in Excel.

Here is an example - it is the statquotes data, from File > Open from Library > Statquotes package
image

I wanted to find a particular quote - the one about `liars`.

This is a common problem.  But it isn't as general as in Excel, because almost always there is a single variable that I need to examine.  So I would be quite happy with a `Find ` that brings up a simple dialogue that just searches in one variable.  
And, of course, we already (sort of) have the dialogue in the Prepare > Column Text > Find/Replace 

image

So I would be quite happy with just a simple copy of this as a new dialogue.  
a) The usual selector on the left 
b) Single receiver on right  (meant for strings, but can take anything)
c) Pattern field
d) Ignore case
e) Include Regular Expressions (is also available.)

It then uses the stringr command to locate the first occurrence of the pattern.
The difference is what it does then.  It should mark that row, or text in some way. and go there. 

It should work on what's visible - so if a filter is in place, then it just searches the visible text.  I think that will be automatic.

Initially I am quite happy if it always starts at the top.  Later we could permit it perhaps to start from a particular line - that's not so trivial when the data are filtered perhaps.  Perhaps it would be easy to have a control 
From (Visible) Row and then an up-down control?  If so, then it could perhaps start on the current row of the data frame, when the dialogue is opened, or re-opened.  

One use of this dialogue is that the help could direct users to the other text options.  Perhaps this Find dialogue  could even have label saying `Use Prepare > Column: Text > Find/Replace for more options`

We could also perhaps implement `Find Next`, fairly trivially.  It would be disabled if there is no Find text.  It is enabled if there is, and there is a short-cut - perhaps <ctrl>,shift>F?  It doesn't give a dialogue, just goes to the next occurrence.

Then I propose we delete the Replace and Cut items from the menu.

### LINKED ISSUE #8427: Bug on the new Edit cell feature
@derekagorhom I found this bug while testing the find in rows/variable I have implemented. This occurs when trying to edit a cell not in the first page of the grid.
image
From this below you can see that this bug occurs because we are using the row header as row index which is wrong.
image
I will fix this in my PR #8426 and you can have a look.

### PR DESCRIPTION
**Pull request by N-thony**
replaces #8423
@rdstern please test.
Fixes #8427
Fixes #7950
I have also added CTRL + F to open the dialgoue.

### CODE REVIEW

**Comment by rdstern** [comment]
@N-thony I checked also the Edit cell and that seems ok.

In your Find dialog it is a bit off that the find control has the TRUE/FALSE as the default options?  Was that always there?

image

**Comment by N-thony** [comment]
> @N-thony I checked also the Edit cell and that seems ok.
> 
> In your Find dialog it is a bit off that the find control has the TRUE/FALSE as the default options? Was that always there?
> 
> image

@rdstern you have requested this in PR #8152 
image

**Comment by rdstern** [comment]
@N-thony on the drop down, so I did!  The surpriuse is that TRUE is the default.  Could the default be blank instead?  Would it also be possible to include NA to get missing as another option?

**Review by rdstern — DISMISSED** [review]
@N-thony nice now.  It is easy to think of improvements that could be added, but this is useful now.  So let's merge as soon as @lloyddewit is ok with the code.

**Comment by N-thony** [comment]
> @N-thony nice now. It is easy to think of improvements that could be added, but this is useful now. So let's merge as soon as @lloyddewit is ok with the code.

@rdstern I was still making some changes in the R function to be able to find rows with NA.

**Comment by lloyddewit** [comment]
@N-thony 
Does this PR also fix #7950? If so, should we add this to the top comment?

**Comment by N-thony** [comment]
> @N-thony Does this PR also fix #7950? If so, should we add this to the top comment?

@lloyddewit that is correct

File: instat/dlgFindInVariableOrFilter.vb [hunk 1]
```diff
@@ -1,3 +1,198 @@
...
+                If lstRowNumbers.Count <= 0 Then
+                    lblMatching.ForeColor = Color.Red
+                    lblMatching.Text = "There are no entries matching " & ucrInputPattern.GetText
+                    lblMatching.Visible = True
+                    Exit Sub
+                End If
+
+                Dim iStartRowHeader As Integer = frmMain.ucrDataViewer.GetFirstRowHeader
+                Dim iEndRowHeader As Integer = frmMain.ucrDataViewer.GetLastRowHeader
+                For i As Integer = 1 To lstRowNumbers.Count
```
Comment (lloyddewit) [inline instat/dlgFindInVariableOrFilter.vb:135]: I cannot easily understand what's happening in this `For` loop, add a comment?

File: instat/dlgFindInVariableOrFilter.vb [hunk 2]
```diff
@@ -1,3 +1,198 @@
...
+                    Dim iRowIndex As Integer = lstRowNumbers(i - 1)
+                    If ((iRowIndex >= iStartRowHeader AndAlso iRowIndex <= iEndRowHeader) OrElse iRowIndex > iEndRowHeader) AndAlso
+                        (CInt(lstRowNumbers(iFisrtRow - 1)) < iRowIndex) Then
```
Comment (lloyddewit) [inline instat/dlgFindInVariableOrFilter.vb:138]: Exceeds 100 columns, split over multiple lines?
Can the condition be simplified?
```suggestion
                    If iRowIndex >= iStartRowHeader \
                        AndAlso CInt(lstRowNumbers(iFisrtRow - 1)) < iRowIndex Then
```

**Review by lloyddewit — CHANGES_REQUESTED** [review]
@N-thony This is nicely implemented and a significant piece of work - thank you.
If you're able to resolve the comments today, then we can include in release 0.7.16 tomorrow.

File: instat/frmMain.vb [hunk 3]
```diff
@@ -915,9 +915,7 @@ Public Class frmMain
     End Sub
 
     Private Sub mnuEditFind_Click(sender As Object, e As EventArgs) Handles mnuEditFind.Click
-        dlgFind.currWindow = ActiveMdiChild
-        dlgFind.Owner = Me
-        dlgFind.Show()
```
[the remark this answers is not in the document]
Comment (N-thony) [reply instat/frmMain.vb:920]: Resolved already

File: instat/dlgFindInVariableOrFilter.vb [hunk 4]
```diff
@@ -1,3 +1,198 @@
...
+' This program is distributed in the hope that it will be useful,
+' but WITHOUT ANY WARRANTY; without even the implied warranty of
+' MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+' GNU General Public License for more details.
+'
+' You should have received a copy of the GNU General Public License 
+' along with this program.  If not, see <http://www.gnu.org/licenses/>.
 
+Imports instat.Translations
+Imports RDotNet
```
[the remark this answers is not in the document]
Comment (N-thony) [reply instat/dlgFindInVariableOrFilter.vb:18]: needed on line 120 where we are executing internally the R code to get the row headers from the corresponfing dataframe.

File: instat/frmMain.vb (hunk 3 above)
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/frmMain.vb:920]: @N-thony the `dlgFind.*` files are still in `master` branch and this PR does not remove these files. Please could you clarify 'resolved already'? thanks

File: instat/dlgFindInVariableOrFilter.vb [hunk 5]
```diff
@@ -1,3 +1,198 @@
...
+            cmdFind.Enabled = True
+        Else
+            cmdFind.Enabled = False
+        End If
+    End Sub
+
+    Private Sub cmdFind_Click(sender As Object, e As EventArgs) Handles cmdFind.Click
+        Try
+            Dim strPattern As String = ucrInputPattern.GetText
+            If strPattern <> "" Then
```
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/dlgFindInVariableOrFilter.vb]: @N-thony Thanks for making this change, please could you also fix the indentation for the rows below?

File: instat/dlgFindInVariableOrFilter.vb (hunk 4 above)
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/dlgFindInVariableOrFilter.vb:18]: yes, you're correct, sorry for that

File: instat/UserControls/DataGrid/ReoGrid/ucrDataViewReoGrid.vb [hunk 6]
```diff
@@ -211,7 +221,91 @@ Public Class ucrDataViewReoGrid
...
+            End If
+        Next
+    End Sub
+
+    Public Sub SearchInGrid(rowNumbers As List(Of String), strColumn As String, Optional iRow As Integer = 0,
+                            Optional bCellOrRow As Boolean = False) Implements IDataViewGrid.SearchInGrid
+        Dim currSheet = grdData.CurrentWorksheet
+
+        If currSheet.RowHeaders.Any(Function(x) x.Text = iRow) Then
+            Dim iRowIndex = GetRowIndex(currSheet, iRow)
```
[the remark this answers is not in the document]
Comment (lloyddewit) [reply instat/UserControls/DataGrid/ReoGrid/ucrDataViewReoGrid.vb]: @N-thony Sorry, I think there's a typo in my suggested code, hopefully the meaning is still clear

**Comment by lloyddewit** [comment]
@N-thony Thank you for making a start with this. I appreciate that you're not finished yet. I reviewed anyway and I think there are just 6 open comments left.
When you work on these comments, please don't set to 'resolved' (or I can't see which comments I still need to review). Just let me know when this PR is ready to review again. Thanks

**Comment by N-thony** [comment]
@lloyddewit I think I resolved your comments.

File: instat/dlgFindInVariableOrFilter.vb (hunk 1 above)
Comment (lloyddewit) [reply instat/dlgFindInVariableOrFilter.vb]: @N-thony Thank you for the comment. I still don't really understand what's going on. Please can we discuss? thanks

**Comment by N-thony** [comment]
@lloyddewit I think I resolved all your comments, please have a look.

**Comment by N-thony** [comment]
> @lloyddewit I think I resolved all your comments, please have a look.

@lloyddewit  I tried with the click counter as index and now I remember that the reason of the for loop was to avoid always starting at the top row. For example if the text is on page 1,2,5,9 and I'm on page 9, if I click ok it will always go back on page 1. @rdstern has suggested that it should start on the current page if the row or text exist. I'm thinking of putting the for loop back.

**Review by rdstern — APPROVED** [review]
@lloyddewit and @N-thony I am still ok with this as a useful feature for now - once the code is solid.

**Review by lloyddewit — CHANGES_REQUESTED** [review]

**Review by lloyddewit — APPROVED** [review]
@N-thony Thanks, looks good, happy to approve.
I just removed some comments that referred to the old variable names. Thank you for your patience

**Review by rdstern — APPROVED** [review]
@lloyddewit and @N-thony it is now better than I thought, as it automatically moves from page to page.  Excellent new feature to merge.

**Pull request merged**